src/rest_api/classes/laboratory/widgets/members.clj

```diff
@@ -1,31 +1,108 @@
+                         (date/format-date5 ((first (json-data role)) rto)))]] to)
+        duration
+        (for [role roles
+              :let [rfrom (keyword (-> role (str/replace #"/" ".") (str/replace #":" "")) "from")
+                    rto (keyword (-> role (str/replace #"/" ".") (str/replace #":" "")) "to")
+                    from (if (contains? (first (json-data role)) rfrom)
+                           (date/format-date5 ((first (json-data role)) rfrom)))
+                    to (if (contains? (first (json-data role)) rto)
+                         (date/format-date5 ((first (json-data role)) rto)))
+                    duration (str from " - " to)]] duration)]
 
-(defn current-members [l]
-  {:data (some->> (:laboratory/registered-lab-members l)
-                  (map get-lineage-data))
-   :description "current members of the laboratory"})
+    { (keyword other-person-id) {:level (first level)
+                                 :person other-person-id
+                                 :duration (first duration)
+                                 :start_date (first from)
+                                 :end_date (first to) } }))
 
-(defn former-members [l]
-  {:data (some->> (:labloratory/past-lab-members l)
-                  (map get-lineage-data))
-   :description "former members of the laboratory"})
+
+(defn- generate-lineage [laboratory]
+  (let [db (d/entity-db laboratory)
+        reps (:laboratory/representative laboratory)
+        lineage
+        (apply merge
+               (flatten 
+                 (for [rep reps] 
+                   (->> (d/q '[:find ?other-person-id ?other-person-name (pull ?person-supervised-by-ent [*])
+                               :in $ ?person
+                               :where
+                               [?other-person-ent :person/supervised-by ?person-supervised-by-ent]
+                               [?person-supervised-by-ent :person.supervised-by/person ?person]
+                               [?other-person-ent :person/id ?other-person-id]
+                               [?other-person-ent :person/standard-name ?other-person-name]]
+                             db (:db/id rep))
+                        (map (fn [oid]
+                               (generate-map oid)))(seq)))))]
+    lineage))
+
+(defn current-members [laboratory]
+  (let [db (d/entity-db laboratory)
+        lineage (generate-lineage laboratory)
+        data (->> (d/q '[:find [?members ...]
+                         :in $ ?laboratory
+                         :where
+                         [?laboratory :laboratory/registered-lab-members ?members]]
+                       db (:db/id laboratory))
+                  (map (fn [members]
+                         (let [member (d/entity db members)
+                               lab (d/entity db laboratory)]
+                           {:name (pack-obj "person" member)
+                            :level (:level ((keyword (:person/id member)) lineage))
```

Comment:
I am not sure what ((keyword (:person/id member)) lineage)) does but you could pull it out into a variable instead of regenerating it for each of these lines.

Reply:
Oh, I see, I didn't think it would be computationally expensive, compared to creating extra variables.  Sure though, changed.